app/services/sso_service.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class SSOService
+  def initialize(saml_response)
+    raise 'SAML Response is required' if saml_response.nil?
+    @saml_response = saml_response
+    @saml_attributes = SAML::User.new(@saml_response)
+  end
+
+  attr_reader :new_session, :new_user, :new_user_identity, :saml_attributes
+
+  def existing_user
+    @existing_user ||= User.find(saml_attributes.user_attributes.uuid)
+  end
+
+  def self.extend!(session, user)
+    session.expire(Session.redis_namespace_ttl)
+    user&.identity&.expire(UserIdentity.redis_namespace_ttl)
+    user&.expire(User.redis_namespace_ttl)
+  end
+
+  def persist_authentication!
+    return false unless @saml_response.is_valid?
+    @new_user_identity = UserIdentity.new(saml_attributes.to_hash)
+    @new_user = init_new_user(new_user_identity, existing_user, saml_attributes.changing_multifactor?)
+    @new_session = Session.new(uuid: new_user.uuid)
+
+    if existing_user.present?
+      existing_user&.identity&.destroy
+      existing_user.destroy
+    end
+
+    @new_session.save && @new_user.save && @new_user_identity.save
+  end
+
+  private
+
+  def expire_existing!
+    existing_user&.identity&.destroy
```

Comment:
could be DRYed up with 29-30 above, and maybe moved into user as a single call?
```ruby
# in user.rb
def obliterate!
  self.identity&.destroy
  self.destroy
  nil
end

@existing_user = existing_user&.obliterate!

existing_user&.obliterate! # removes need for .present? check
```

Reply:
Similarly here. Do you think we really need another destructive method? If anything I would just override the existing destroy logic maybe since I can't think of a situation where you would need UserIdentity without a User (which does beg the question why it was separated in the first place I guess, but thats nuanced ;))

To the point earlier though, do we plan on expiring a session anywhere other than Logout or Persisting a new one (leveling up / multifactor)? If so I think it maybe shouldn't exist on the model then. Again, I can see it either way. Maybe further refactor will warrant revisiting?